### PR TITLE
[TRTLLM-13349][perf] Fuse gemma RMSNorm into AllReduce for Qwen3-Next/Qwen3.5…

### PR DESCRIPTION
**Pull request by nv-guomingz**
… TEP

Enable the eager AllReduce + RMSNorm fusion on the Qwen3-Next/Qwen3.5 full-attention decoder layers in the tensor-parallel (non attention_dp) path, using a gemma-aware fused kernel.

image

## Summary by CodeRabbit

* **New Features**
  * Added support for Gemma-style RMSNorm scaling in fused all-reduce operations, controllable via environment variable configuration.

* **Refactor**
  * Improved weight handling alignment between fused and non-fused computation paths.
  * Enhanced tensor-parallel fusion configuration logic for improved MoE support.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: cpp/tensorrt_llm/thop/allreduceOp.cpp [hunk 1]
```diff
@@ -684,6 +685,10 @@ class AllreduceOp
         allreduce_fusion_params.residual_out = nullptr;
         allreduce_fusion_params.norm_out = nullptr;
         allreduce_fusion_params.trigger_completion_at_end = trigger_completion_at_end;
+        // Opt-in gemma-style RMSNorm (scale by 1+gamma) inside the fused AR+RMSNorm
+        // kernel. Off by default so behavior is identical to baseline for all other
+        // models; enabled for Qwen3-Next/Qwen3.5 (all-gemma norms) via env var.
+        allreduce_fusion_params.use_gemma = tensorrt_llm::common::getBoolEnv("TLLM_QWEN3_AR_FUSION_GEMMA_NORM");
```
[the remark this answers is not in the document]
Comment (nv-guomingz) [reply cpp/tensorrt_llm/thop/allreduceOp.cpp]: resolved

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 2]
```diff
@@ -60,6 +60,22 @@
...
+    fused AR+RMSNorm kernels (and the NCCL / NCCL_SYMMETRIC fallbacks the
+    AUTO strategy may pick) only apply ``weight``. Baking the ``+1`` into the
+    weight here makes EVERY allreduce backend produce the correct gemma result
+    without any backend-specific flag. The ``+1`` is done in fp32 then cast
+    back, matching the unfused gemma path.
+    """
+    w = norm.weight
+    if getattr(norm, "use_gemma", False):
+        return (w.float() + 1.0).to(w.dtype)
+    return w
```
[the remark this answers is not in the document]
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:86]: resolved

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 3]
```diff
@@ -366,17 +411,15 @@ def __init__(
...
-        self.enable_fusion &= not self.enable_attention_dp
+        self.enable_fusion = _eager_fusion_enabled(self.enable_attention_dp)
 
         has_tp = self.mapping.has_tp()
         has_pp = self.mapping.has_pp()
 
         self.fusion_config.PRE_MOE_FUSION = self.enable_fusion and has_tp
-        self.fusion_config.POST_MOE_FUSION = self.fusion_config.PRE_MOE_FUSION and not has_pp and self.enable_attention_dp
-        self.disable_attn_allreduce = (self.mapping.tp_size == 1
+        self.fusion_config.POST_MOE_FUSION = self.fusion_config.PRE_MOE_FUSION and not has_pp
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py:429]: ~Attention DP gate removed, intentional?~ Looks like the comment below explains this

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 4]
```diff
@@ -531,17 +572,26 @@ def __init__(self, model_config: ModelConfig[Qwen3NextConfig],
...
 
-        self.fusion_config.POST_MOE_FUSION = self.fusion_config.PRE_MOE_FUSION and not has_pp and self.enable_attention_dp
-        self.disable_attn_allreduce = (self.mapping.tp_size == 1
+        # POST_MOE_FUSION fuses the MoE-output all-reduce with the next layer's
+        # RMSNorm. It is a tensor-parallel (TEP) optimization: it is only valid
+        # when ranks share the same tokens (not attention_dp, where each rank holds
+        # different tokens and the MoE block does no cross-rank all-reduce). This
+        # mirrors the DeepSeek-V3 pattern (POST == PRE in the non-attention_dp path).
+        # NOTE: the prior `and self.enable_attention_dp` made this always False
+        # (PRE already implies not-attention_dp), silently disabling the fusion.
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: This NOTE should probably be removed, it won't make much sense outside of the context of this PR

**Review by mikeiovine — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 4 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Updated

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 5]
```diff
@@ -62,6 +62,58 @@
 from .modeling_utils import DecoderModel, EagerFusionConfig, register_auto_model
 
 
+def _fused_norm_weight(norm: RMSNorm) -> torch.Tensor:
+    """Weight to feed the fused AllReduce+RMSNorm op for ``norm``.
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py:66]: Nit: switch from double to single backticks on changed lines.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 6]
```diff
@@ -62,6 +62,58 @@
...
+
+    Gemma RMSNorm scales by ``(1 + weight)`` (see RMSNorm.forward), but the
+    fused AR+RMSNorm kernels (and the NCCL / NCCL_SYMMETRIC fallbacks the
+    AUTO strategy may pick) only apply ``weight``. Baking the ``+1`` into the
+    weight makes EVERY allreduce backend produce the correct gemma result
+    without any backend-specific flag.
+
+    For gemma norms the ``(1 + weight)`` tensor is precomputed once in
+    :meth:`Qwen3NextForCausalLM.cache_derived_state` and cached on the module as
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py]: Nit: `:meth:` can be removed.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 7]
```diff
@@ -62,6 +62,58 @@
...
+    ``_fused_norm_weight``. Computing it inline here would re-run a cast+add
+    elementwise kernel every forward inside the CUDA graph. The inline path
+    below is only a correctness fallback if the cache is absent.
+    """
+    cached = getattr(norm, "_fused_norm_weight", None)
+    if cached is not None:
+        return cached
+    w = norm.weight
+    if getattr(norm, "use_gemma", False):
+        return (w.float() + 1.0).to(w.dtype)
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py:85]: For my own benefit, why the `float()`?

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py [hunk 8]
```diff
@@ -1016,3 +1076,7 @@ def setup_aliases(self) -> None:
             else:
                 layer.next_layer_layernorm = self.model.layers[
                     idx + 1].input_layernorm
+
+    def cache_derived_state(self) -> None:
+        super().cache_derived_state()
+        _precompute_fused_norm_weights(self)
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_qwen3_next.py:1082]: Maybe stupid question / just making sure I understand - could the same effect have been achieved if we had a load state dict hook that modified the `weight` parameters for `RMSNorm` layers?

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    assert model.model.layers[0].next_layer_layernorm is model.model.layers[1].input_layernorm
+    assert model.model.layers[1].next_layer_layernorm is model.model.norm
+    assert not hasattr(model.model.norm, "_fused_norm_weight")
+
+
+def test_cache_derived_state_refreshes_gemma_norm_weight() -> None:
+    model = _new_causal_lm()
+    model.gemma_norm = RMSNorm(hidden_size=4, eps=1e-6, dtype=torch.bfloat16, use_gemma=True)
+    model.standard_norm = RMSNorm(hidden_size=4, eps=1e-6, dtype=torch.bfloat16)
+    with torch.no_grad():
```
Comment (2ez4bz) [inline tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: Just curious, is there a reason not to have all the test functions in here be decorated with this? Or `inference_mode`?

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    expected = (model.gemma_norm.weight.float() + 1.0).to(torch.bfloat16)
+    torch.testing.assert_close(model.gemma_norm._fused_norm_weight, expected)
+    assert not hasattr(model.standard_norm, "_fused_norm_weight")
+
+    with torch.no_grad():
+        model.gemma_norm.weight.add_(1.0)
+    model.cache_derived_state()
+    expected = (model.gemma_norm.weight.float() + 1.0).to(torch.bfloat16)
+    torch.testing.assert_close(model.gemma_norm._fused_norm_weight, expected)
```
Comment (2ez4bz) [inline tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: Should it be closeness or equality we check for?

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    assert not _eager_fusion_enabled(enable_attention_dp=True)
+
+    monkeypatch.setenv("TRTLLM_QWEN3_EAGER_FUSION_DISABLED", "1")
+    assert not _eager_fusion_enabled(enable_attention_dp=False)
+
+
+def test_gdn_fusion_has_single_allreduce_owner() -> None:
+    hidden_states = torch.randn(2, 4, dtype=torch.bfloat16)
+    residual = torch.randn_like(hidden_states)
+    post_attention_norm = SimpleNamespace(
```
Comment (2ez4bz) [inline tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: Do we need to use `SimpleNamespace` for these simpler layer types? In general I think it's better for the test if we can use the actual objects being tested rather than mocks.

Same for pretty much all the "layer" instantiations below.

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+        position_ids=torch.arange(2),
+        hidden_states=hidden_states,
+        attn_metadata=SimpleNamespace(),
+        residual=residual,
+    )
+
+    internal_ar_params = linear_attn.call_args.kwargs["all_reduce_params"]
+    assert not internal_ar_params.enable_allreduce
+
+    assert allreduce.call_count == 2
```
Comment (2ez4bz) [inline tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py:144]: Nit: leave a comment for why this is the expected value?

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    pre_ar_params = allreduce.call_args_list[0].kwargs["all_reduce_params"]
+    post_ar_params = allreduce.call_args_list[1].kwargs["all_reduce_params"]
+    assert pre_ar_params.enable_allreduce
+    assert post_ar_params.enable_allreduce
+    assert pre_ar_params.fusion_op == AllReduceFusionOp.RESIDUAL_RMS_NORM
+    assert post_ar_params.fusion_op == AllReduceFusionOp.RESIDUAL_RMS_NORM
+    torch.testing.assert_close(
+        pre_ar_params.norm_weight,
+        (post_attention_norm.weight.float() + 1.0).to(torch.bfloat16),
```
Comment (2ez4bz) [inline tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py:154]: Similar question as above re closeness vs equality.

**Review by 2ez4bz — COMMENTED** [review]
Approving modeling side changes.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 8 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:1082]: Good question. A load hook that rewrites norm.weight to (1+weight) would corrupt the non-fused path: RMSNorm.forward still applies its own gemma (1 + weight), so a pre-baked weight would double-count (1 + (1+weight)). You'd then also have to change forward to stop adding 1 — coupling this optimization into the standard norm path and every non-fused backend. The current approach leaves norm.weight untouched (normal forward unchanged) and caches the derived (1+weight) as a separate _fused_norm_weight attribute consumed only by the fused AR+RMSNorm backend. Cleaner separation, and it's opt-in per backend.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 7 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:85]: w is bf16 and gemma weights sit near 0; w + 1.0 in bf16 rounds the small w against 1.0's exponent and drops low mantissa bits. Doing the add in fp32 and rounding to bf16 once yields the correctly-rounded 1+w, matching gemma RMSNorm's fp32 (1 + weight) semantics.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 5 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:66]: This file consistently uses double backticks (RST inline-literal) for code refs — 15 such lines vs 3 single. Keeping double here matches the file convention; single backticks in RST render as title-reference (italic), not literal. Happy to switch the whole file in a separate cleanup if preferred.

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py (hunk 10 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: Agreed — expected recomputes _fused_norm_weight via the identical fp32-add-then-cast that cache_derived_state uses, so it's deterministic and must be bitwise-equal. Switched to exact equality (atol=rtol=0) so any divergence in the baking is caught rather than masked by tolerance.

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py (hunk 12 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py:144]: updated

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py (hunk 11 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: Good point — switched the norm stand-ins to real RMSNorm objects (they're lightweight and their actual weight/use_gemma semantics are what we're exercising). Kept linear_attn/allreduce as Mock()s deliberately: this test asserts on how they're called (call_count/call_args for the single-allreduce-owner behavior), which needs mocks, not real objects.

Comment (2ez4bz) [reply tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py]: We could spy on them to have both real objects and `call_count` / `call_args` available.

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 5 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:66]: Backticks are claude / codex produced artifacts. Our docs are not RST-based AFAICT so they are just a waste of space, hence why I suggested switching from double to single backticks only on changed lines (to not include unrelated changes into the PR).

File: tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py (hunk 13 above)
Comment (nv-guomingz) [reply tests/unittest/_torch/models/test_qwen3_next_eager_fusion.py:154]: Updated

**Review by zongfeijing — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 5 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py:66]: Let's format the code entirely next time

File: tensorrt_llm/_torch/models/modeling_qwen3_next.py (hunk 6 above)
Comment (nv-guomingz) [reply tensorrt_llm/_torch/models/modeling_qwen3_next.py]: updated

**Review by Wanli-Jiang — APPROVED** [review]
